Writer and WriterT documentation site pages (#3618)
typelevel/cats: docs/src/main/mdoc/datatypes/writer.md

---
layout: docs
title:  "Writer"
section: "data"
source: "core/src/main/scala/cats/data/package.scala"
scaladoc: "#cats.data.Writer"
---
# Writer

> The `Writer[L, A]` datatype represents a computation that produce a

Reviewer: ```suggestion
The `Writer[L, A]` datatype represents a computation that produces a
```

tuple containing a value of type `L` and one of type `A`. Usually, the
value `L` represents a description of the computation. A typical
example of an `L` value could be a logging `String` and that's why
from now on we will refer to it as the _Logging side_ of the
datatype. Meanwhile, the value `A` is the actual output of the
computation.

> The main features the `Writer` provides are:

Reviewer: ```suggestion
The main features that `Writer` provides are:
```

- The flexibility regarding Log value management. It can be modified
in multiple ways. See the [Operations section](#operations)
- When two functions are composed together, eg using `flatMap`, the

>   logs of both functions will be combined using an implicit [Semigroup](https://typelevel.org/cats/typeclasses/semigroup.html).

Reviewer: Both are good points, but if it's an enumeration, the structure of the sentences should be similar. I suggest removing the bullet points and rephrasing it as a paragraph.

## Operations

> The `Writer` datatype provides a set of functions that are quite

Reviewer: "quite identical" → "similar to"

identical to the ones from the [Monad](https://typelevel.org/cats/typeclasses/monad.html) typeclass. In fact, they share

> the same name and the same signature, but for the requirement of a

Reviewer: ```suggestion
the same name and the same signature, but have an additional requirement of a
```

[`Semigroup[L]`](https://typelevel.org/cats/typeclasses/semigroup.html) that allows the log merging.

> `map` effects only the value, keeping the log side untouched. Plus, here we show `run`

Reviewer: ```suggestion
`map` affects only the value, keeping the log side untouched. Plus, here we show `run`
```
